java/ql/test/experimental/query-tests/security/CWE-555/CredentialsInPropertiesFile.ql

```diff
@@ -0,0 +1,84 @@
+/**
```

Comment:
Shouldn't this file be a `.qlref` file referencing the query, instead of duplicating its code?

Reply:
A <code>.qlref</code> file doesn't work in this case. The reason is that to reduce FPs, there is a predicate <code>isNonProdCredentials(CredentialsConfig cc)</code> that checks whether the properties file is in a directory with the name <code>dev</code>, <code>test</code>, or <code>sample</code>; or has a file name with those words in it. 

However, as all CodeQL test cases are in the folder <code>test/experimental/query-tests/security/CWE-xxx/</code>, this predicate excludes expected results from being included.

Is there a better way to model this? Thanks.

Reply:
Ah that makes sense. Though duplicating the query seems pretty error-prone, it is likely that in the future only one of them is adjusted by accident.
@smowton, do you have an idea how this can be improved?

Reply:
Maybe the explicit check for `test` in the file name could also be omitted because LGTM by default shows no alerts for [files classified as test](https://lgtm.com/help/lgtm/file-classification); though that likely does not cover complete integration test projects where the complete project is only used for testing.

Reply:
@Marcono1234 LGTM doesn't seem to be able to detect test properties files by default. If I change <code>isNonProdCredentials</code> to:
```
predicate isNonProdCredentials(CredentialsConfig cc) {
  cc.getFile().getAbsolutePath().matches(["%dev%", "%sample%"])
}
``` 
Then test properties files like <code>swif-testing.properties</code> in the referenced sample project will be included in the query results.

Reply:
You could put most of the query in a .qll file that is imported by a .ql file in src and executed, and imported by a .ql file in test and filtered.

Reply:
Thanks @smowton for the advice. I've created a separate <code>.qll</code> file to reuse the code. I also added code to check the most common usage pattern of properties file with Spring.